service/collector/rate_limit.go

```diff
@@ -0,0 +1,262 @@
+	HostAzureClientSetConfig client.AzureClientSetConfig
+}
+
+func NewRateLimit(config RateLimitConfig) (*RateLimit, error) {
+	if config.G8sClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.G8sClient must not be empty", config)
+	}
+	if config.K8sClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.K8sClient must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	if config.EnvironmentName == "" {
+		return nil, microerror.Maskf(invalidConfigError, "%T.EnvironmentName must not be empty", config)
+	}
+	if config.Location == "" {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Location must not be empty", config)
+	}
+
+	u := &RateLimit{
+		g8sClient: config.G8sClient,
+		k8sClient: config.K8sClient,
+		logger:    config.Logger,
+
+		environmentName:          config.EnvironmentName,
+		location:                 config.Location,
+		HostAzureClientSetConfig: config.HostAzureClientSetConfig,
+	}
+
+	return u, nil
+}
+
+func (u *RateLimit) Collect(ch chan<- prometheus.Metric) error {
+	// We need all CRs to gather all subscriptions below.
+	var crs []providerv1alpha1.AzureConfig
+	{
+		mark := ""
+		page := 0
+		for page == 0 || len(mark) > 0 {
+			opts := metav1.ListOptions{
+				Continue: mark,
+			}
+			list, err := u.g8sClient.ProviderV1alpha1().AzureConfigs(metav1.NamespaceAll).List(opts)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+
+			crs = append(crs, list.Items...)
+
+			mark = list.Continue
+			page++
+		}
+	}
+
+	// We generate clients and group them by subscription.
+	// Subscription is defined by the Secret used to fetch the credentials.
+	clientConfigBySubscription := map[string]client.AzureClientSetConfig{}
+	{
+		for _, cr := range crs {
+			config, err := credential.GetAzureConfig(u.k8sClient, key.CredentialName(cr), key.CredentialNamespace(cr))
+			if err != nil {
+				return microerror.Mask(err)
+			}
+			clientConfigBySubscription[config.SubscriptionID] = *config
+		}
+
+		if _, present := clientConfigBySubscription[u.HostAzureClientSetConfig.SubscriptionID]; !present {
+			config := &client.AzureClientSetConfig{
+				ClientID:       u.HostAzureClientSetConfig.ClientID,
+				ClientSecret:   u.HostAzureClientSetConfig.ClientSecret,
+				SubscriptionID: u.HostAzureClientSetConfig.SubscriptionID,
+				TenantID:       u.HostAzureClientSetConfig.TenantID,
+			}
+			clientConfigBySubscription[config.SubscriptionID] = *config
+		}
+	}
+
+	ctx := context.Background()
+
+	// We track RateLimit metrics for each client labeled by SubscriptionID and
+	// ClientID.
+	// That way we prevent duplicated metrics.
+	for _, clientConfig := range clientConfigBySubscription {
+		azureClients, err := client.NewAzureClientSet(clientConfig)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		resourceGroup, err := u.createEmptyResourceGroup(ctx, azureClients)
```

Comment:
When creating randomly named RG on each Collect it's likely to have an orphaned RG. E.g. when we hit rate limits.

I would prefer having some RG with a fixed name and never delete it. I assume you can take write limit from  `azureClients.GroupsClient.CreateOrUpdate` call.

Reply:
If this works, then +1.

Reply:
Done